addons/website/models/website_visitor.py

```diff
@@ -166,88 +187,96 @@ def action_send_mail(self):
             'context': compose_ctx,
         }
 
-    def _get_visitor_from_request(self, force_create=False):
+    def _get_visitor_from_request(self, force_create=False, force_track_values=None):
         """ Return the visitor as sudo from the request if there is a
-        visitor_uuid cookie.
-
-        When fetching visitor, now that duplicates are linked to a main visitor
-        instead of unlinked, you may have more collisions issues with cookie
-        being set (after a de-connection for example).
+        visitor for that device-ip.
 
-        The visitor associated to a partner in case of public user is not taken
-        into account, it is considered as desynchronized cookie.
-        In addition, we also discard if the visitor has a main visitor whose
-        partner is set (aka wrong after logout partner). """
+        :param bool force_create: force a visitor creation if no visitor exists
+        :param force_track_values: an optional dict to create a track at the
+            same time.
+        :return: the website visitor if exists or forced, empty recordset
+                 otherwise.
+        """
 
         # This function can be called in json with mobile app.
         # In case of mobile app, no uid is set on the jsonRequest env.
         # In case of multi db, _env is None on request, and request.env unbound.
-        if not request:
+        if not (request and request.env and request.env.uid):
             return None
+
         Visitor = self.env['website.visitor'].sudo()
         visitor = Visitor
-        access_token = request.httprequest.cookies.get('visitor_uuid')
-        if access_token:
-            visitor = Visitor.with_context(active_test=False).search([('access_token', '=', access_token)])
-            # Prefetch access_token and other fields. Since access_token has a restricted group and we access
-            # a non restricted field (partner_id) first it is not fetched and will require an additional query to be retrieved.
-            visitor.access_token
+        access_token = self._get_access_token()
+
+        if force_create:
+            # This will create or return existing one
+            visitor_create_values = Visitor._get_visitor_create_values()
+            if len(str(access_token)) != 40:
+                visitor_create_values['partner_id'] = access_token
+            visitor_field_names = ', '.join(visitor_create_values.keys())
+            visitor_field_values_format_placeholder = ', '.join('%s' for _ in visitor_create_values)
+            query = f"""
+                INSERT INTO website_visitor ({visitor_field_names})
+                VALUES ({visitor_field_values_format_placeholder})
+                ON CONFLICT (access_token)
+                DO UPDATE SET
+                    last_connection_datetime='{visitor_create_values['last_connection_datetime']}',
```

Comment:
that's an SQLi vector. Not sure why you would not pass this in the `query_values`, as this time it was not even an SQL identifier?

Reply:
You can also simply use the excluded value
```suggestion
                    last_connection_datetime=excluded.last_connection_datetime,
```